hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelInboundPipeline.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.networking;
+
+/**
+ *
+ * <h1>Spurious wakeups</h1>
+ * ChannelInboundHandlers/ChannelOutboundHandlers need to be able to deal with
+ * spurious wakeups. E.g. it could be that a PacketDecoder is called with a null
+ * Packet.
+ *
+ * todo: rationale behind this
+ *
+ * <h1>Automatic reprocessing on change</h1>
+ * When a change in the pipeline is detected, the pipeline is automatically
+ * reprocessed. For example when the ProtocolDecoder replaced itself by a
+ * PacketDecoder, the whole pipeline (in this case the PacketDecoder) is
+ * automatically reprocessed.
+ */
+public interface ChannelInboundPipeline {
+
+    /**
+     * Adds the handlers at the end of the pipeline.
+     *
+     * No verification is done if the handler is already added and a handler
+     * should only be added once.
+     *
+     * This method should only be made on the thread 'owning' the handler.
+     *
+     * @param handlers the handlers to add
+     * @return this
+     */
+    ChannelInboundPipeline addLast(ChannelInboundHandler... handlers);
+
+      /**
+     * Replaces the old ChannelInboundHandler by the new ones.
```

Comment:
Does it replace in the same order? 
For example, assume there are 3 handlers in order: `H1, H2, H3`.  
What will be the handlers, when `replace(H2, H4, H5)` is called? `H1, H4, H5, H3` ?

Reply:
Good point; let me add it to the documentation.